### PR TITLE
refactor(key_manager): integrate key manager in vault apis

### PR DESCRIPTION
**Pull request by Chethan-rao**
This PR adds support for providing key manager functionality, both externally and internally and uses feature flag to enable/disable them

(With `external_key_manager` feature flag enabled) 

1. Add card

image

Entity table - 

image

2. Retrieve card

image

3. Delete card

image

(With `external_key_manager` feature flag disabled)

4. When `master_key` is invalid

image

image

1. Add card

image

Merchant table - 

image

2. Retrieve card

image

3. Delete card

image

### CODE REVIEW

File: src/crypto/keymanager/internal_keymanager.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    ) -> Result<Box<dyn CryptoOperationsManager>, ContainerError<error::ApiError>> {
+        let master_encryption =
+            GcmAes256::new(tenant_app_state.config.tenant_secrets.master_key.clone());
+
+        Ok(tenant_app_state
+            .db
+            .find_by_merchant_id(&entity_id, &master_encryption)
+            .await
+            .map(|inner| InternalCryptoManager::from_secret_key(inner.enc_key))
+            .map(|inner| -> Box<dyn CryptoOperationsManager> { Box::new(inner) })?)
```
Comment (NishantJoshi00) [inline src/crypto/keymanager/internal_keymanager.rs:30]:
```suggestion
            .map(Box::new)?)
```

File: src/crypto/keymanager/internal_keymanager.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            GcmAes256::new(tenant_app_state.config.tenant_secrets.master_key.clone());
+
+        let entity = tenant_app_state
+            .db
+            .find_or_create_by_merchant_id(&entity_id, &master_encryption)
+            .await;
+
+        let response = entity
+            .map(|inner| InternalCryptoManager::from_secret_key(inner.enc_key))
+            .map(|inner| -> Box<dyn CryptoOperationsManager> { Box::new(inner) })?;
```
Comment (NishantJoshi00) [inline src/crypto/keymanager/internal_keymanager.rs:48]: same here

File: src/routes/data/crypto_operation.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+pub async fn decrypt_data(
+    tenant_app_state: &TenantAppState,
+    crypto_operator: Box<dyn CryptoOperationsManager>,
+    mut locker: Locker,
+) -> Result<Locker, ContainerError<error::ApiError>> {
+    let decrypted_data = crypto_operator
+        .decrypt_data(tenant_app_state, locker.enc_data)
+        .await?;
+
+    locker.enc_data = decrypted_data;
```
Comment (NishantJoshi00) [inline src/routes/data/crypto_operation.rs:47]: This increases the changes of data inconsistency. I am worried that if this function isn't called we might deal with encrypted data everywhere and we wouldn't know. Worst case the opposite might happen.

Should we consider investing some time on using something like.

```rust
enum Encryptable {
    Encrypted(Secret<Vec<u8>>),
    Plaintext(StrongSecret<String>)
}
```

File: src/storage.rs [hunk 4]
```diff
@@ -137,13 +139,17 @@ pub(crate) trait MerchantInterface {
...
     /// Insert a new merchant in the database by encrypting the dek with `master_key`
     async fn insert_merchant(
         &self,
         new: types::MerchantNew<'_>,
         key: &Self::Algorithm,
     ) -> Result<types::Merchant, ContainerError<Self::Error>>;
 
+    // This function is under the `dead_code` lint to pass Clippy checks because it utilizes types
+    // from both internal and external key_manager.
+    #[allow(dead_code)]
```
Comment (NishantJoshi00) [inline src/storage.rs:152]: Why dead code though? can the lint be handled with `cfg` feature flags?

File: src/storage/caching.rs [hunk 5]
```diff
@@ -5,18 +5,31 @@ use super::types;
...
+pub trait CacheableWithEntity<T>: super::Cacheable<types::Entity> {}
+
+#[cfg(feature = "external_key_manager")]
+impl<T: super::Cacheable<types::Entity>> CacheableWithEntity<T> for T {}
+
+#[cfg(not(feature = "external_key_manager"))]
+pub trait CacheableWithEntity<T> {}
+
+#[cfg(not(feature = "external_key_manager"))]
+impl<T> CacheableWithEntity<T> for T {}
```
Comment (NishantJoshi00) [inline src/storage/caching.rs:18]: Nice one dude!

File: src/storage/consts.rs [hunk 6]
```diff
@@ -12,4 +12,5 @@ pub const ID_LENGTH: usize = 20;
 /// Header Constants
 pub mod headers {
     pub const CONTENT_TYPE: &str = "Content-Type";
+    pub const AUTHORIZATION: &str = "Authorization";
```
Comment (NishantJoshi00) [inline src/storage/consts.rs:15]: funny enough both of these might not be required.

- [`AUTHORIZATION`](https://docs.rs/hyper/latest/hyper/header/constant.AUTHORIZATION.html)
- [`CONTENT_TYPE`](https://docs.rs/hyper/latest/hyper/header/constant.CONTENT_TYPE.html)

File: src/storage/db.rs [hunk 7]
```diff
@@ -146,15 +149,14 @@ impl LockerInterface for Storage {
             })
             .map_err(error::ContainerError::from)
             .map_err(From::from)
-            .and_then(|inner| Ok(inner.decrypt(key)?))
+            .map(|inner| inner.into())
```
Comment (NishantJoshi00) [inline src/storage/db.rs:152]:
```
.map_or_else(From::from, From::from)
```
This is just a nit, not something to worry over

File: src/storage/utils.rs [hunk 8]
```diff
@@ -1,5 +1,10 @@
 use crate::storage::consts;
 
-pub fn generate_id(id_length: usize) -> String {
+pub fn generate_nano_id(id_length: usize) -> String {
     nanoid::nanoid!(id_length, &consts::ALPHABETS)
 }
+
+/// Generate UUID v4 as strings to be used in storage layer
+pub fn generate_uuid() -> String {
+    uuid::Uuid::new_v4().to_string()
```
Comment (NishantJoshi00) [inline src/storage/utils.rs:9]: Should we invest efforts in going with [`UUIDv7`](https://uuid7.com/)

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

**Review by NishantJoshi00 — APPROVED** [review]
Looks good, can you please attach some screenshots as the changes in the key manager are merged.

**Review by NishantJoshi00 — APPROVED** [review]

**Review by ShankarSinghC — APPROVED** [review]

**Pull request merged**